### ISSUE TITLE
ASS subtitles break if rendering can't keep up

### ISSUE THREAD

**Issue by Streetwalrus** [mpv-player/mpv#6286]
### mpv version and platform

Arch Linux, 3770k, RX 580
```
mpv 0.29.1 Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on Fri Oct  5 08:08:09 CEST 2018
ffmpeg library versions:
   libavutil       56.14.100
   libavcodec      58.18.100
   libavformat     58.12.100
   libswscale      5.1.100
   libavfilter     7.16.100
   libswresample   3.1.100
ffmpeg version: n4.0.3
```

### Reproduction steps

Play a file with large ASS signs, at a high resolution (I can reproduce this on my 4k monitor in full screen but not in a smaller ~1080p window).

### Expected behavior

Sign renders correctly, even if that would cause stutter (or skip only the offending sign).

### Actual behavior

Video stutters for a split second, subtitles stop rendering for the remainder of this mpv session.

### Notes

The most reliable workaround I've found for now is to use `blend-subtitles=video`, since that forces rendering at the video's lower resolution, but that isn't ideal.
Seeking past the sign in question before it shows up prevents the problem from being triggered.

### Log file

I didn't see anything of interest in the log but I can provide one if absolutely necessary.

### Sample files

`[Asenshi-GJM-PAS] SSSS.GRIDMAN - 04 [518F25EE].mkv` at around 8:50 (not sure if it's okay to post this here, will remove if asked)

**Comment by Streetwalrus** [comment]
I'm messing with the code a little, trying to find the root cause. It looks like `mp_image_params_valid` rejects the dimensions. When I try to render this sign in full screen, it tries to allocate a 16384x16384 image, which sounds ridiculously large. Even at the video's native 720p, it still tries to allocate an 8192x4096 image, which seems just as ridiculous to me.

**Comment by Streetwalrus** [comment]
Alright, I've figured out exactly what happens:
* libass is properly rendering the sign at the requested resolution, but it has a lot of layers
* with the libass bitmap format, mpv tries to pack those layers into an atlas, which grows in size very quickly with window resolution
* this causes the atlas to reach the 16k*16k limit and allocation fails
* mpv gets stuck trying to allocate a buffer of the same size on each frame, so subs stop rendering entirely (possibly another bug that is being tripped as a side effect?)
* forcing RGBA mode bypasses the problem because the packer for this format blits the entire layer stack in software instead of making an atlas (which explains the massive performance drop)

I would gladly submit a patch for this, but I have absolutely no idea how to approach this since the check is buried deep into a long chain of functions. The comment also makes it pretty clear that it's not really a good idea to remove it:
https://github.com/mpv-player/mpv/blob/8b6064de76ce5472a7d00b3eb7c4e38d01754330/video/mp_image.c#L662-L670
Thoughts?

**Comment by haasn** [comment]
I mean the basic approach would be to split up the atlas into multiple atlases (atli?)

**Comment by Streetwalrus** [comment]
That sounds reasonable enough to me, but it also looks like a pretty big refactor (all VO drivers would have to be updated). I'll investigate how feasible it is tomorrow.

> (atli?)

Atlas is Greek, so definitely not. :P Looks like the correct plural is "atlantes".

**Comment by Streetwalrus** [comment]
Just looked into this some more, looks like subtitles break completely afterwards because the packer struct is stuck with the oversized dimensions. This lets it continue gracefully once the subs shrink back to a more normal size.

```diff
--- a/sub/ass_mp.c
+++ b/sub/ass_mp.c
@@ -224,8 +224,10 @@ static bool pack(struct mp_ass_packer *p, struct sub_bitmaps *res, int imgfmt)
     {
         talloc_free(p->cached_img);
         p->cached_img = mp_image_alloc(imgfmt, p->packer->w, p->packer->h);
-        if (!p->cached_img)
+        if (!p->cached_img) {
+            packer_reset(p->packer);
             return false;
+        }
         talloc_steal(p, p->cached_img);
     }
 
```

It'd be helpful to log a warning in that case, since it will still cause the subtitles to disappear when oversized, but adding that in looks awfully convoluted.

**Comment by wm4** [comment]
Yes, this is a hard problem. I think vsfilter just rendered directly into the video, and so didn't have this issue. I wonder if newer vsfilter forks did better.

Maybe the actual solution for this is to let libass create large, pre-blended RGBA overlays.

If you have a screen larger than the video resolution, enabling the option to render subtitles in video resolution might help.

All in all probably a too obscure use case to put too much effort into.

@Streetwalrus can you post a git forma-patch (or a PR) for this patch you posted?

**Comment by Streetwalrus** [comment]
> Maybe the actual solution for this is to let libass create large, pre-blended RGBA overlays.

As long as it doesn't incur the same performance impact as `force-rgba-osd-rendering` does, that would be the best solution I suppose.

> If you have a screen larger than the video resolution, enabling the option to render subtitles in video resolution might help.

Yeah, I've been using `blend-subtitles=video` which works fine.

> @Streetwalrus can you post a git forma-patch (or a PR) for this patch you posted?

I'll prepare a PR later.

**Comment by frozenpandaman** [comment]
`blend-subtitles=video` isn't working for me; doesn't change anything about the playback of this file which still hangs/skips (and doesn't render any subtitles) during frames with the large ASS signs…

**Comment by wm4** [comment]
Maybe the supported texture size is smaller for you. And rendering it seems to be very slow anyway, even if you don't hit the texture limit described in this issue.

@9ary: reading this issue again, I think the only VO you'd have to change is vo_gpu, and maybe the software rendering fallback code. Some other VOs support textured ASS overlays, but barely anyone uses them, so they could just fail as usual in this case.

On the other hand, trying to handle this case better in libass would probably give better results (if possible).

**Comment by 9ary** [comment]
> Maybe the supported texture size is smaller for you. And rendering it seems to be very slow anyway, even if you don't hit the texture limit described in this issue.

I doubt it, it's a hardcoded constant. It's possible that they've run into a sample that's somehow even worse though.

> @9ary: reading this issue again, I think the only VO you'd have to change is vo_gpu, and maybe the software rendering fallback code. Some other VOs support textured ASS overlays, but barely anyone uses them, so they could just fail as usual in this case.

That sounds acceptable to me. I think my concern back then was probably that the different graphics APIs were still separate VO backends. I also lack the setup and knowledge to test this on DirectX. That said, maybe it's possible to relax the rules specifically for ASS overlays. Are those ever passed to libavfilter?

**Comment by wm4** [comment]
I mean he could get the "OSD bitmaps do not fit on a surface with "... message, which would also lead to no subs rendered.

vo_gpu has an abstraction layer that's on a quite low level, so you don't have to care whether the underlying API is GL, Vulkan, or d3d (much of the portable code still has "gl" in its function and type names for legacy reasons). vo_direct3d is a legacy thing which I meant you should ignore. In cases where multiple textures are needed, VOs using the "old" way would simply get no subs. The new way could be a flag to osd_draw (or whatever else works).

**Comment by wm4** [comment]
Another idea: maybe it could fall back to a RGBA overlay on overflow (there is code to render libass output to an RGBA overlay).

Regarding your libavfilter comment: no, mpv's rendered subtitles never interact with ffmpeg.

**Comment by CounterPillow** [comment]
Since it has been four years, I thought I should mention that this is still reproducible on current git master, though since I don't have access to a 4K monitor I reproduced as follows:

```
mpv --no-config -vf gpu=w=3840:h=2160
```

I don't know how applicable that is to e.g. gpu-next.

**Comment by 9ary** [comment]
As far as I remember this is completely independent from the VO.
